### PR TITLE
cmd/alertmanager: reject invalid external URLs

### LINKED ISSUE #1912: Template email not replacing external and generator URL
**What did you do?**
In the compose I set up:
command:
- '--web.external-url=localhost:9093'
But still in the email I am getting docker ID and the port, which I guess is default behavior.

**What did you expect to see?**
I want my email to have correct Alert Manager and source addresses.

* Alertmanager version:

	Latest from docker

* Prometheus version:

	Latest from docker

### PR DESCRIPTION
**Pull request by simonpasquier**
Closes #1912

Using `--web.external-url=foo:123` instead of `--web.external-url=http://foo:123` was accepted before but it's (almost) never what you want as `html/template` will escape the value to `#ZgotmplZ` because it only accepts `http://...`, `https://...` or `mailto://...` links.

### CODE REVIEW

File: cmd/alertmanager/main.go [hunk 1]
```diff
@@ -481,16 +482,25 @@ func clusterWait(p *cluster.Peer, timeout time.Duration) func() time.Duration {
...
+type hostnameFunc func() (string, error)
+
+func extURL(logger log.Logger, listen, external string) (*url.URL, error) {
+	return hostnameFunc(os.Hostname).resolveExtURL(logger, listen, external)
+}
+
+func (h hostnameFunc) resolveExtURL(logger log.Logger, listen, external string) (*url.URL, error) {
 	if external == "" {
-		hostname, err := os.Hostname()
+		hostname, err := h()
```
Comment (brian-brazil) [inline cmd/alertmanager/main.go]: Where's this defined?

File: api/api.go [hunk 2]
```diff
@@ -96,7 +96,7 @@ func (o Options) validate() error {
 // call is also needed to get the APIs into an operational state.
 func New(opts Options) (*API, error) {
 	if err := opts.validate(); err != nil {
-		return nil, fmt.Errorf("invalid API options: %s", err)
+		return nil, errors.Wrap(err, "invalid API options")
```
Comment (brian-brazil) [inline api/api.go]: There seems to be unrelated changes in here.

File: cmd/alertmanager/main.go [hunk 3]
```diff
@@ -499,7 +509,11 @@ func extURL(listen, external string) (*url.URL, error) {
 	if err != nil {
 		return nil, err
 	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return nil, errors.Errorf("%q: invalid %q scheme, only 'http' and 'https' are supported", u.String(), u.Scheme)
+	}
 
+	fmt.Println("path:", u.Path, "raw", u.RawPath)
```
Comment (brian-brazil) [inline cmd/alertmanager/main.go]: Debug code?

File: api/api.go (hunk 2 above)
Comment (simonpasquier) [reply api/api.go]: Yup, will do it in another PR.

File: cmd/alertmanager/main.go (hunk 3 above)
Comment (simonpasquier) [reply cmd/alertmanager/main.go]: Totally :)

File: cmd/alertmanager/main.go (hunk 1 above)
Comment (simonpasquier) [reply cmd/alertmanager/main.go]: It's the method receiver.

**Comment by brian-brazil** [comment]
:+1:

File: cmd/alertmanager/main.go [hunk 4]
```diff
@@ -481,16 +482,25 @@ func clusterWait(p *cluster.Peer, timeout time.Duration) func() time.Duration {
 	}
 }
 
-func extURL(listen, external string) (*url.URL, error) {
+type hostnameFunc func() (string, error)
```
Comment (mxinden) [inline cmd/alertmanager/main.go]: What is this indirection through `extURL -> hostnameFunc -> resolveExtURL` for? I don't understand why this is not just a single function.

Comment (simonpasquier) [reply cmd/alertmanager/main.go]: To resolve the external URL we need to get the hostname but I didn't want to call `os.Hostname` in the tests so I went for this convoluted way. I'll add a comment unless you can think of something clearer.

Comment (mxinden) [reply cmd/alertmanager/main.go]: How about passing `os.Hostname` as an argument to `resolveExtURL`? Would remove the complexity of instantiating a new type.
